src/wallet/wallet_rpc_server.cpp

```diff
@@ -198,24 +199,68 @@ namespace tools
+
+      try 
+      {
         bool received_money = false;
-        if (m_wallet) m_wallet->refresh(m_wallet->is_trusted_daemon(), 0, blocks_fetched, received_money, true, true, REFRESH_INFICATIVE_BLOCK_CHUNK_SIZE);
-      } catch (const std::exception& ex) {
+        if (m_wallet) m_wallet->refresh(m_wallet->is_trusted_daemon(), 0, blocks_fetched, received_money, true, true, REFRESH_INDICATIVE_BLOCK_CHUNK_SIZE);
+        refresh_success = true;
+      } 
+      catch (const std::exception& ex)
+      {
         LOG_ERROR("Exception at while refreshing, what=" << ex.what());
       }
-      // if we got the max amount of blocks, do not set the last refresh time, we did only part of the refresh and will
-      // continue asap, and only set the last refresh time once the refresh is actually finished
-      if (blocks_fetched < REFRESH_INFICATIVE_BLOCK_CHUNK_SIZE)
-        m_last_auto_refresh_time = boost::posix_time::microsec_clock::universal_time();
+
+      if (blocks_fetched == 0)
+        return true;
+
+      const auto end = std::chrono::system_clock::now();
+      const auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+      if (refresh_success)
+        LOG_PRINT_L3((boost::format(tr("Automated wallet block refresh took %i ms")) % elapsed.count()).str());
+
+      const bool syncing_against_tip_of_chain = blocks_fetched < REFRESH_INDICATIVE_BLOCK_CHUNK_SIZE;
+      if (syncing_against_tip_of_chain)
+      {
+        // At this point, we can poll for a refresh every m_auto_refresh_period seconds.
+        m_last_auto_refresh_time = end;
+      }
+      else
+      {
+        // We are in a state of synchronization, blasting through the maximum chunks of blocks
+        // because we are not at the tip of the chain. In this case, if we update m_last_auto_refresh_time,
+        // we'll need to wait an entire m_refresh_interval_ms before processing the next batch. On the other hand, 
+        // if we do not update m_last_auto_refresh_time, we'll never yield (other calls to the RPC will hang)
+        // in the case that elapsed > auto_refresh_evaluation_ms since we'll immediately be scheduled for another block sync.
+        const bool over_one_refresh_period_passed = end > m_last_auto_refresh_time + auto_refresh_interval_ms;
+        if (over_one_refresh_period_passed)
+        {
+          // auto_refresh_interval_ms of straight-blasting through blocks has elapsed without end.
+          // Let's wait till proceeding for at least 3 timeout intervals, freeing us up to respond to other requests in that time.
+          const auto refresh_throttle = auto_refresh_evaluation_ms * 3;
+          m_last_auto_refresh_time = end - auto_refresh_interval_ms + refresh_throttle;
+          LOG_PRINT_L3((boost::format(tr("Temporarily throttling wallet block refresh by %i ms")) % refresh_throttle.count()).str());
+        }
+      }
       return true;
-    }, 1000);
+    }, auto_refresh_evaluation_ms.count());
```

Comment:
just a question: 
Should this one be evaluation or interval?

Reply:
It's the evaluation interval. Interval presumes that the refresh is guaranteed to happen, but all the 1000ms guarantees is that we'll test to see if the `refresh` is valid.